distribution/lib/Standard/Base/0.0.0-dev/src/Data/Statistics.enso

```diff
@@ -139,44 +153,70 @@ type Statistic
+                statistics.map s-> case s of
+                    Statistic.Count -> counter.count
+                    Statistic.Minimum -> if counter.comparatorError then (Error.throw Incomparable_Values) else counter.minimum
+                    Statistic.Maximum -> if counter.comparatorError then (Error.throw Incomparable_Values) else counter.maximum
+                    Statistic.Covariance s -> calculate_correlation_statistics data s . covariance
+                    Statistic.Pearson s -> calculate_correlation_statistics data s . pearsonCorrelation
+                    Statistic.Spearman s -> calculate_spearman_rank data s
+                    Statistic.R_Squared s -> calculate_correlation_statistics data s . rSquared
+                    _ -> moments.compute (to_moment_statistic s)
+
+    ## Compute a running statistics on a vector like object.
+
+       Arguments:
+       - data: Vector like object which has a `to_array` method.
+       - statistics: Set of statistics to calculate.
+    running : Vector -> Statistic -> Vector Any
+    running data statistic=Statistic.Sum =
+        Statistic.running_bulk data [statistic] . map .first
+
+    ## Compute a set running statistics on a vector like object.
+
+       Arguments:
+       - data: Vector like object which has a `to_array` method.
+       - statistics: Set of statistics to calculate.
+    running_bulk : Vector -> Vector Statistic -> Vector Any
+    running_bulk data statistics=[Statistic.Count, Statistic.Sum] =
+        unsupported s = case s of
+            Statistic.Covariance _ -> True
+            Statistic.Pearson _ -> True
+            Statistic.Spearman _ -> True
+            Statistic.R_Squared _ -> True
+            _ -> False
+        if statistics.any unsupported then Error.throw (Illegal_Argument.Error ("Unsupported Statistics ( " + (statistics.filter unsupported . to_text) ") for running calculations.")) else
+            counter = CountMinMax.new
+
+            java_stats = statistics.map .order
+            skip_moments = java_stats.all s->s.is_nothing
+            moments = if skip_moments then Nothing else Moments.new (java_stats.filter v->v!=Nothing . fold 0 .max)
+            ref_moment = Ref.new moments
+
+            data.running_fold Nothing _->v->
+                result = compute_fold [counter, ref_moment.get] v
+                ref_moment.put result.second
+
+                statistics.map s-> case s of
+                    Statistic.Count -> counter.count
+                    Statistic.Minimum -> if counter.comparatorError then (Error.throw Incomparable_Values) else counter.minimum
+                    Statistic.Maximum -> if counter.comparatorError then (Error.throw Incomparable_Values) else counter.maximum
+                    _ -> if result.second.is_nothing then Error.throw (Illegal_Argument.Error ("Can only compute " + s.to_text + " on numerical data sets.")) else result.second.compute (to_moment_statistic s)
```

Comment:
I don't see why we are using `running_fold` if we are discarding the accumulator.
Why not use `each` with `Vector.Builder`? That would at least make the stateful aspect of this computation clearer - now it is pretty hidden in that Ref.